### DISCUSSION TITLE
better support for typescript in subscribe?

### DISCUSSION THREAD

**Discussion by chrisdrackett** [pmndrs/zustand#600] [Q&A]
I'm curious if anyone has a way to get better types for the state within the subscribe function. Seems to me that this should work similar to the hook where the selector you provide helps type what is returned, but at least in my experience the state is only just `object`.

**Comment by dai-shi** [comment]
It turns out it's not possible with the current API and the current TypeScript. #555

A workaround for you is to create a helper function.
```ts
const subscribeWithSelector = <T extends State, U>(
  store: StoreApi<T>,
  selector: (state: T) => U,
  listener: (slice: U) => void
) => store.subscribe(listener, selector);
```

See the working example: https://codesandbox.io/s/inspiring-babycat-vmucu?file=/src/App.tsx
